Incorrectly returned INVALID for ropsten block

### Description
On one of our ropsten nodes, besu incorrectly returned `INVALID` with a 0x0000 latestValidHash for beacon block 957088 (root 0x1cfaadb29cab42bb951aca0d068f624cbf4ae68fe5117c8c133114a84b96fb6f)

Execution payload header:
```
        "execution_payload": {
          "parent_hash": "0x9abbfb6c82d7c0b9639333acd2d648205c12c16db0e304877ec765a822323803",
          "fee_recipient": "0x965b7a773e3632b259108d246a7cfdcdff118999",
          "state_root": "0x93489b3724363a97c28808016d360e630ffe99d3449c8f275163dcb1120652ab",
          "receipts_root": "0xc97efdeae4ddcbb597099852fe158644f86a6411c8db862c4f698526474980dd",
          "logs_bloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
          "prev_randao": "0xcf371ca53a22567031cb5c909a1c9f27f417240b795a718c9bcb269efdea8325",
          "block_number": "13125902",
          "gas_limit": "30000000",
          "gas_used": "346036",
          "timestamp": "1665407856",
          "extra_data": "0x",
          "base_fee_per_gas": "7",
          "block_hash": "0x56128e3165fa54e3644b2de9493972741a029666f128b9b2595a0df47be61d5f",
```

This caused the node to be unable to follow the canonical chain.  The block was correctly imported after restarting besu and then teku.

Relevant teku logs:
```

{"@timestamp":"2022-10-10T13:17:36,270","level":"INFO","thread":"beaconchain-async-0","class":"teku-status-log","message":"Sync committee performance: epoch 29908, expected 768, produced 768, correct 768, included 672 (88%)","throwable":""}
{"@timestamp":"2022-10-10T13:17:37,124","level":"ERROR","thread":"pool-4-thread-417","class":"teku-event-log","message":"Execution Client request failed. Make sure the Execution Client is online and in sync.","throwable":" java.io.IOException: -32603: Internal error\n\tat tech.pegasys.teku.ethereum.executionclient.web3j.Web3JClient.lambda$doRequest$0(Web3JClient.java:85)\n\tat java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)\n\tat java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:148)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)\n\tat org.web3j.utils.Async.lambda$run$1(Async.java:38)\n\tat java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)\n\tat java.base/java.lang.Thread.run(Thread.java:833)\n"}
{"@timestamp":"2022-10-10T13:17:37,134","level":"ERROR","thread":"forkchoice-async-0","class":"BlockManager","message":"Unable to import block: Execution Client returned an error: tech.pegasys.teku.ethereum.executionlayer.InvalidRemoteResponseException: Invalid remote response from the execution client: -32603: Internal error","throwable":""}
{"@timestamp":"2022-10-10T13:17:37,136","level":"ERROR","thread":"beaconchain-async-4","class":"teku-validator-log","message":"Failed to import proposed block due to FAILED_EXECUTION_PAYLOAD_EXECUTION: 1cfaadb29cab42bb951aca0d068f624cbf4ae68fe5117c8c133114a84b96fb6f (957088)","throwable":" java.util.concurrent.CompletionException: tech.pegasys.teku.ethereum.executionlayer.InvalidRemoteResponseException: Invalid remote response from the execution client: -32603: Internal error\n\tat java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)\n\tat java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)\n\tat java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:649)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:148)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:148)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:887)\n\tat java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2325)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:603)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:33)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:143)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$exceptionallyCompose$33(SafeFuture.java:422)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:148)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)\n\tat org.web3j.utils.Async.lambda$run$1(Async.java:38)\n\tat java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)\n\tat java.base/java.lang.Thread.run(Thread.java:833)\nCaused by: tech.pegasys.teku.ethereum.executionlayer.InvalidRemoteResponseException: Invalid remote response from the execution client: -32603: Internal error\n\tat tech.pegasys.teku.ethereum.executionlayer.ExecutionLayerManagerImpl.unwrapResponseOrThrow(ExecutionLayerManagerImpl.java:518)\n\tat tech.pegasys.teku.ethereum.executionlayer.ExecutionLayerManagerImpl.unwrapExecutionClientResponseOrThrow(ExecutionLayerManagerImpl.java:504)\n\tat java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)\n\t... 29 more\n"}
{"@timestamp":"2022-10-10T13:17:37,647","level":"INFO","thread":"validator-async-2","class":"teku-validator-log","message":"Validator   *** Published block              Count: 1, Slot: 957088, Root: 1cfaadb29cab42bb951aca0d068f624cbf4ae68fe5117c8c133114a84b96fb6f, 346036 (1%) gas, EL block:  56128e3165fa54e3644b2de9493972741a029666f128b9b2595a0df47be61d5f (13125902)","throwable":""}
{"@timestamp":"2022-10-10T13:17:39,135","level":"INFO","thread":"beaconchain-async-1","class":"FailedExecutionPool","message":"Retrying execution of block 1cfaadb29cab42bb951aca0d068f624cbf4ae68fe5117c8c133114a84b96fb6f (957088)","throwable":""}
{"@timestamp":"2022-10-10T13:17:39,139","level":"INFO","thread":"pool-4-thread-417","class":"teku-event-log","message":"Execution Client request succeeded after a previous failure","throwable":""}
{"@timestamp":"2022-10-10T13:17:39,169","level":"WARN","thread":"forkchoice-async-0","class":"ForkChoiceStrategy","message":"Payload for child of block root 0xd85399eb10de6804a2d8d5d749a457b350a6c2bbedc3858dc24be275072d3c63 marked as invalid by Execution Client","throwable":""}
{"@timestamp":"2022-10-10T13:17:39,170","level":"ERROR","thread":"forkchoice-async-0","class":"teku-status-log","message":"Could not find latest valid execution payload hash (0x0000000000000000000000000000000000000000000000000000000000000000) in the non-finalized chain. Optimistic sync may have finalized an invalid transition block.","throwable":""}
```

Besu logs:
```
{"@timestamp":"2022-10-10T13:17:24,224","level":"INFO","thread":"vert.x-worker-thread-0","class":"EngineNewPayload","message":"Imported #13,125,901 / 0 tx / base fee 7 wei / 0 (0.0%) gas / (0x9abbfb6c82d7c0b9639333
acd2d648205c12c16db0e304877ec765a822323803) in 0.001s. Peers: 25","throwable":""}
{"@timestamp":"2022-10-10T13:17:24,910","level":"INFO","thread":"EthScheduler-Computation-534","class":"MergeCoordinator","message":"Successfully built block 13125902 (0x56128e3165fa54e3644b2de9493972741a029666f128
b9b2595a0df47be61d5f) for proposal identified by 0x0035eb5652a22a90, with 5 transactions, in 108ms","throwable":""}
{"@timestamp":"2022-10-10T13:17:24,999","level":"INFO","thread":"EthScheduler-Computation-534","class":"MergeCoordinator","message":"Successfully built block 13125902 (0x56128e3165fa54e3644b2de9493972741a029666f128
b9b2595a0df47be61d5f) for proposal identified by 0x0035eb5652a22a90, with 5 transactions, in 188ms","throwable":""}
{"@timestamp":"2022-10-10T13:17:36,043","level":"INFO","thread":"vert.x-worker-thread-0","class":"EngineGetPayload","message":"Fetch block proposal by identifier: 0x0035eb5652a22a90, hash: 0x56128e3165fa54e3644b2de
9493972741a029666f128b9b2595a0df47be61d5f, number: 13125902, coinbase: 0x965b7a773e3632b259108d246a7cfdcdff118999, transaction count: 5","throwable":""}
{"@timestamp":"2022-10-10T13:17:37,092","level":"ERROR","thread":"vert.x-worker-thread-0","class":"AbstractBlockProcessor","message":"failed persisting block","throwable":" org.hyperledger.besu.plugin.services.exce
ption.StorageException: org.rocksdb.RocksDBException: TimedOut(LockTimeout)\n\tat org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage$RocksDbTransaction.put(RocksDBColumnar
KeyValueStorage.java:299)\n\tat org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage$RocksDbTransaction.put(RocksDBColumnarKeyValueStorage.java:280)\n\tat org.hyperledger.be
su.services.kvstore.SegmentedKeyValueStorageTransactionTransitionValidatorDecorator.put(SegmentedKeyValueStorageTransactionTransitionValidatorDecorator.java:36)\n\tat org.hyperledger.besu.services.kvstore.Segmented
KeyValueStorageAdapter$1.put(SegmentedKeyValueStorageAdapter.java:80)\n\tat org.hyperledger.besu.ethereum.bonsai.BonsaiWorldStateKeyValueStorage$Updater.putAccountInfoState(BonsaiWorldStateKeyValueStorage.java:318)
\n\tat org.hyperledger.besu.ethereum.bonsai.BonsaiPersistedWorldState.calculateRootHash(BonsaiPersistedWorldState.java:223)\n\tat org.hyperledger.besu.ethereum.bonsai.BonsaiInMemoryWorldState.rootHash(BonsaiInMemor
yWorldState.java:43)\n\tat org.hyperledger.besu.ethereum.bonsai.BonsaiInMemoryWorldState.persist(BonsaiInMemoryWorldState.java:53)\n\tat org.hyperledger.besu.ethereum.mainnet.AbstractBlockProcessor.processBlock(Abs
tractBlockProcessor.java:220)\n\tat org.hyperledger.besu.ethereum.mainnet.AbstractBlockProcessor.processBlock(AbstractBlockProcessor.java:43)\n\tat org.hyperledger.besu.ethereum.mainnet.BlockProcessor.processBlock(
BlockProcessor.java:80)\n\tat org.hyperledger.besu.ethereum.MainnetBlockValidator.processBlock(MainnetBlockValidator.java:175)\n\tat org.hyperledger.besu.ethereum.MainnetBlockValidator.validateAndProcessBlock(Mainn
etBlockValidator.java:112)\n\tat org.hyperledger.besu.consensus.merge.blockcreation.MergeCoordinator.validateBlock(MergeCoordinator.java:343)\n\tat org.hyperledger.besu.consensus.merge.blockcreation.MergeCoordinator.rememberBlock(MergeCoordinator.java:359)\n\tat org.hyperledger.besu.consensus.merge.blockcreation.TransitionCoordinator.rememberBlock(TransitionCoordinator.java:137)\n\tat org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods.engine.EngineNewPayload.syncResponse(EngineNewPayload.java:215)\n\tat org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods.ExecutionEngineJsonRpcMethod.lambda$response$0(ExecutionEngineJsonRpcMethod.java:73)\n\tat io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:159)\n\tat io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)\n\tat io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:157)\n\tat io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)\n\tat io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)\n\tat java.base/java.lang.Thread.run(Thread.java:833)\nCaused by: org.rocksdb.RocksDBException: TimedOut(LockTimeout)\n\tat org.rocksdb.Transaction.put(Native Method)\n\tat org.rocksdb.Transaction.put(Transaction.java:705)\n\tat org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage$RocksDbTransaction.put(RocksDBColumnarKeyValueStorage.java:293)\n\t... 25 more\n"}
{"@timestamp":"2022-10-10T13:17:37,115","level":"ERROR","thread":"vert.x-worker-thread-0","class":"MainnetBlockValidator","message":"Error processing block. Block 13125902 (0x56128e3165fa54e3644b2de9493972741a029666f128b9b2595a0df47be61d5f), caused by Optional[org.hyperledger.besu.plugin.services.exception.StorageException: org.rocksdb.RocksDBException: TimedOut(LockTimeout)]","throwable":""}
{"@timestamp":"2022-10-10T13:17:37,178","level":"INFO","thread":"EthScheduler-Computation-535","class":"MergeCoordinator","message":"Successfully built block 13125902 (0x56128e3165fa54e3644b2de9493972741a029666f128b9b2595a0df47be61d5f) for proposal identified by 0x0035eb5652a22a90, with 5 transactions, in 1146ms","throwable":""}
{"@timestamp":"2022-10-10T13:17:40,024","level":"INFO","thread":"vert.x-worker-thread-0","class":"EngineForkchoiceUpdated","message":"VALID for fork-choice-update: head: 0x9abbfb6c82d7c0b9639333acd2d648205c12c16db0e304877ec765a822323803, finalized: 0x8ba9bc3f1eae2be2dd6f72d0fca0b92539a8d9df918f12c237a0270111aae8b3, safeBlockHash: 0xe6e2279003ab00660ab571b64dee3d82f8550078af727f0b29db0eb1134bc71c","throwable":""}
{"@timestamp":"2022-10-10T13:17:40,092","level":"INFO","thread":"EthScheduler-Computation-536","class":"MergeCoordinator","message":"Successfully built block 13125902 (0x8f679c1cc8e936bb60706a6437df2cc9492613e7f92a98f5036130c962c93580) for proposal identified by 0x0035eb5652a22a84, with 5 transactions, in 68ms","throwable":""}
{"@timestamp":"2022-10-10T13:17:48,028","level":"INFO","thread":"vert.x-worker-thread-0","class":"EngineGetPayload","message":"Fetch block proposal by identifier: 0x0035eb5652a22a84, hash: 0x8f679c1cc8e936bb60706a6437df2cc9492613e7f92a98f5036130c962c93580, number: 13125902, coinbase: 0x965b7a773e3632b259108d246a7cfdcdff118999, transaction count: 5","throwable":""}
{"@timestamp":"2022-10-10T13:17:48,057","level":"INFO","thread":"vert.x-worker-thread-0","class":"EngineNewPayload","message":"Imported #13,125,902 / 5 tx / base fee 7 wei / 346,036 (1.2%) gas / (0x8f679c1cc8e936bb60706a6437df2cc9492613e7f92a98f5036130c962c93580) in 0.004s. Peers: 25","throwable":""}
{"@timestamp":"2022-10-10T13:19:04,015","level":"INFO","thread":"vert.x-worker-thread-0","class":"EngineForkchoiceUpdated","message":"VALID for fork-choice-update: head: 0x8f679c1cc8e936bb60706a6437df2cc9492613e7f92a98f5036130c962c93580, finalized: 0x8ba9bc3f1eae2be2dd6f72d0fca0b92539a8d9df918f12c237a0270111aae8b3, safeBlockHash: 0xe6e2279003ab00660ab571b64dee3d82f8550078af727f0b29db0eb1134bc71c","throwable":""}
{"@timestamp":"2022-10-10T13:20:16,019","level":"INFO","thread":"vert.x-worker-thread-0","class":"EngineForkchoiceUpdated","message":"VALID for fork-choice-update: head: 0x8f679c1cc8e936bb60706a6437df2cc9492613e7f92a98f5036130c962c93580, finalized: 0x8ba9bc3f1eae2be2dd6f72d0fca0b92539a8d9df918f12c237a0270111aae8b3, safeBlockHash: 0xe6e2279003ab00660ab571b64dee3d82f8550078af727f0b29db0eb1134bc71c","throwable":""}
{"@timestamp":"2022-10-10T13:20:16,091","level":"INFO","thread":"EthScheduler-Computation-537","class":"MergeCoordinator","message":"Successfully built block 13125903 (0x3c93b52ee42013c33f3f8a2dfb4e9e7350dd3a099864a3ffdb284c1ca86f6b94) for proposal identified by 0x0035e5100fe7b1e8, with 3 transactions, in 72ms","throwable":""}
```

Initially Besu hit an error and correctly returned an error response (not invalid).  Then Teku retried executing the block and got the incorrect `INVALID` response and from there it went off creating its own fork.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -21,6 +21,7 @@
 - Corrects emission of blockadded events when rewinding during a re-org. Fix for [#4495](https://github.com/hyperledger/besu/issues/4495)
 - Always return a transaction type for pending transactions [#4364](https://github.com/hyperledger/besu/pull/4364)
 - Avoid a cyclic reference while printing EngineExchangeTransitionConfigurationParameter [#4357](https://github.com/hyperledger/besu/pull/4357)
+- Corrects treating a block as bad on internal error [#4512](https://github.com/hyperledger/besu/issues/4512)
 
 ### Download Links
 

--- a/ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayload.java
+++ b/ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayload.java
@@ -107,6 +107,7 @@ public class EngineNewPayload extends ExecutionEngineJsonRpcMethod {
           reqId,
           blockParam,
           mergeCoordinator.getLatestValidAncestor(blockParam.getParentHash()).orElse(null),
+          INVALID,
           "Failed to decode transactions from block parameter");
     }
 
@@ -115,6 +116,7 @@ public class EngineNewPayload extends ExecutionEngineJsonRpcMethod {
           reqId,
           blockParam,
           mergeCoordinator.getLatestValidAncestor(blockParam.getParentHash()).orElse(null),
+          INVALID,
           "Field extraData must not be null");
     }
 
@@ -141,11 +143,12 @@ public class EngineNewPayload extends ExecutionEngineJsonRpcMethod {
     // ensure the block hash matches the blockParam hash
     // this must be done before any other check
     if (!newBlockHeader.getHash().equals(blockParam.getBlockHash())) {
-      LOG.debug(
+      String errorMessage =
           String.format(
               "Computed block hash %s does not match block hash parameter %s",
-              newBlockHeader.getBlockHash(), blockParam.getBlockHash()));
-      return respondWith(reqId, blockParam, null, INVALID_BLOCK_HASH);
+              newBlockHeader.getBlockHash(), blockParam.getBlockHash());
+      LOG.debug(errorMessage);
+      return respondWithInvalid(reqId, blockParam, null, INVALID_BLOCK_HASH, errorMessage);
     }
     // do we already have this payload
     if (protocolContext.getBlockchain().getBlockByHash(newBlockHeader.getBlockHash()).isPresent()) {
@@ -153,13 +156,14 @@ public class EngineNewPayload extends ExecutionEngineJsonRpcMethod {
       return respondWith(reqId, blockParam, blockParam.getBlockHash(), VALID);
     }
     if (mergeCoordinator.isBadBlock(blockParam.getBlockHash())) {
-      return respondWith(
+      return respondWithInvalid(
           reqId,
           blockParam,
           mergeCoordinator
               .getLatestValidHashOfBadBlock(blockParam.getBlockHash())
               .orElse(Hash.ZERO),
-          INVALID);
+          INVALID,
+          "Block already present in bad block manager.");
     }
 
     Optional<BlockHeader> parentHeader =
@@ -170,6 +174,7 @@ public class EngineNewPayload extends ExecutionEngineJsonRpcMethod {
           reqId,
           blockParam,
           mergeCoordinator.getLatestValidAncestor(blockParam.getParentHash()).orElse(null),
+          INVALID,
           "block timestamp not greater than parent");
     }
 
@@ -200,6 +205,7 @@ public class EngineNewPayload extends ExecutionEngineJsonRpcMethod {
           reqId,
           blockParam,
           Hash.ZERO,
+          INVALID,
           newBlockHeader.getHash() + " did not descend from terminal block");
     }
 
@@ -228,7 +234,11 @@ public class EngineNewPayload extends ExecutionEngineJsonRpcMethod {
       }
       LOG.debug("New payload is invalid: {}", executionResult.errorMessage.get());
       return respondWithInvalid(
-          reqId, blockParam, latestValidAncestor.get(), executionResult.errorMessage.get());
+          reqId,
+          blockParam,
+          latestValidAncestor.get(),
+          INVALID,
+          executionResult.errorMessage.get());
     }
   }
 
@@ -237,6 +247,10 @@ public class EngineNewPayload extends ExecutionEngineJsonRpcMethod {
       final EnginePayloadParameter param,
       final Hash latestValidHash,
       final EngineStatus status) {
+    if (INVALID.equals(status) || INVALID_BLOCK_HASH.equals(status)) {
+      throw new IllegalArgumentException(
+          "Don't call respondWith() with invalid status of " + status.toString());
+    }
     debugLambda(
         LOG,
         "New payload: number: {}, hash: {}, parentHash: {}, latestValidHash: {}, status: {}",
@@ -256,7 +270,12 @@ public class EngineNewPayload extends ExecutionEngineJsonRpcMethod {
       final Object requestId,
       final EnginePayloadParameter param,
       final Hash latestValidHash,
+      final EngineStatus invalidStatus,
       final String validationError) {
+    if (!INVALID.equals(invalidStatus) && !INVALID_BLOCK_HASH.equals(invalidStatus)) {
+      throw new IllegalArgumentException(
+          "Don't call respondWithInvalid() with non-invalid status of " + invalidStatus.toString());
+    }
     final String invalidBlockLogMessage =
         String.format(
             "Invalid new payload: number: %s, hash: %s, parentHash: %s, latestValidHash: %s, status: %s, validationError: %s",
@@ -264,7 +283,7 @@ public class EngineNewPayload extends ExecutionEngineJsonRpcMethod {
             param.getBlockHash(),
             param.getParentHash(),
             latestValidHash == null ? null : latestValidHash.toHexString(),
-            INVALID.name(),
+            invalidStatus.name(),
             validationError);
     // always log invalid at DEBUG
     LOG.debug(invalidBlockLogMessage);
@@ -275,7 +294,8 @@ public class EngineNewPayload extends ExecutionEngineJsonRpcMethod {
     }
     return new JsonRpcSuccessResponse(
         requestId,
-        new EnginePayloadStatusResult(INVALID, latestValidHash, Optional.of(validationError)));
+        new EnginePayloadStatusResult(
+            invalidStatus, latestValidHash, Optional.of(validationError)));
   }
 
   private void logImportedBlockInfo(final Block block, final double timeInS) {

--- a/ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadTest.java
+++ b/ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadTest.java
@@ -235,6 +235,8 @@ public class EngineNewPayloadTest {
 
     fromErrorResp(resp);
     verify(engineCallListener, times(1)).executionEngineCalled();
+    verify(mergeCoordinator, times(0)).addBadBlock(any());
+    // verify mainnetBlockValidator does not add to bad block manager
   }
 
   @Test
@@ -362,7 +364,7 @@ public class EngineNewPayloadTest {
     EnginePayloadStatusResult res = fromSuccessResp(resp);
     assertThat(res.getLatestValidHash()).contains(Hash.ZERO);
     assertThat(res.getStatusAsString()).isEqualTo(INVALID.name());
-    assertThat(res.getError()).isNull();
+    assertThat(res.getError()).isEqualTo("Block already present in bad block manager.");
     verify(engineCallListener, times(1)).executionEngineCalled();
   }
 

--- a/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/MainnetBlockValidator.java
+++ b/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/MainnetBlockValidator.java
@@ -151,7 +151,9 @@ public class MainnetBlockValidator implements BlockValidator {
     if (result.causedBy().isPresent()) {
       LOG.info("{}. Block {}, caused by {}", reason, invalidBlock.toLogString(), result.causedBy());
       // TODO: if it's an internal error, don't add it
-      badBlockManager.addBadBlock(invalidBlock);
+      if (!result.internalError()) {
+        badBlockManager.addBadBlock(invalidBlock);
+      }
       return new Result(reason, result.causedBy().get());
     } else {
       LOG.info("{}. Block {}", reason, invalidBlock.toLogString());

--- a/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/MainnetBlockValidator.java
+++ b/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/MainnetBlockValidator.java
@@ -150,7 +150,6 @@ public class MainnetBlockValidator implements BlockValidator {
       final Block invalidBlock, final String reason, final BlockProcessor.Result result) {
     if (result.causedBy().isPresent()) {
       LOG.info("{}. Block {}, caused by {}", reason, invalidBlock.toLogString(), result.causedBy());
-      // TODO: if it's an internal error, don't add it
       if (!result.internalError()) {
         badBlockManager.addBadBlock(invalidBlock);
       }

--- a/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/BlockProcessor.java
+++ b/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/BlockProcessor.java
@@ -22,6 +22,7 @@ import org.hyperledger.besu.ethereum.core.MutableWorldState;
 import org.hyperledger.besu.ethereum.core.Transaction;
 import org.hyperledger.besu.ethereum.core.TransactionReceipt;
 import org.hyperledger.besu.ethereum.privacy.storage.PrivateMetadataUpdater;
+import org.hyperledger.besu.plugin.services.exception.StorageException;
 
 import java.util.List;
 import java.util.Optional;
@@ -64,6 +65,17 @@ public interface BlockProcessor {
 
     default Optional<Throwable> causedBy() {
       return Optional.empty();
+    }
+
+    default boolean internalError() {
+      if (causedBy().isPresent()) {
+        Throwable t = causedBy().get();
+        // As new "internal only" types of exception are discovered, add them here.
+        if (t instanceof StorageException) {
+          return true;
+        }
+      }
+      return false;
     }
   }
 

--- a/ethereum/core/src/test/java/org/hyperledger/besu/ethereum/MainnetBlockValidatorTest.java
+++ b/ethereum/core/src/test/java/org/hyperledger/besu/ethereum/MainnetBlockValidatorTest.java
@@ -36,6 +36,7 @@ import org.hyperledger.besu.ethereum.mainnet.BlockProcessor;
 import org.hyperledger.besu.ethereum.mainnet.HeaderValidationMode;
 import org.hyperledger.besu.ethereum.mainnet.MainnetBlockHeaderFunctions;
 import org.hyperledger.besu.ethereum.worldstate.WorldStateArchive;
+import org.hyperledger.besu.plugin.services.exception.StorageException;
 
 import java.util.Collections;
 import java.util.List;
@@ -239,6 +240,59 @@ public class MainnetBlockValidatorTest {
               @Override
               public boolean isSuccessful() {
                 return true;
+              }
+            });
+    when(blockBodyValidator.validateBody(
+            eq(protocolContext),
+            eq(badBlock),
+            any(),
+            any(),
+            eq(HeaderValidationMode.DETACHED_ONLY)))
+        .thenReturn(true);
+    assertThat(badBlockManager.getBadBlocks().size()).isEqualTo(0);
+    mainnetBlockValidator.validateAndProcessBlock(
+        protocolContext,
+        badBlock,
+        HeaderValidationMode.DETACHED_ONLY,
+        HeaderValidationMode.DETACHED_ONLY);
+    assertThat(badBlockManager.getBadBlocks()).isEmpty();
+  }
+
+  @Test
+  public void shouldNotCacheWhenInternalError() {
+    when(blockchain.getBlockHeader(any(Hash.class)))
+        .thenReturn(Optional.of(new BlockHeaderTestFixture().buildHeader()));
+    when(blockHeaderValidator.validateHeader(
+            any(BlockHeader.class),
+            any(BlockHeader.class),
+            eq(protocolContext),
+            eq(HeaderValidationMode.DETACHED_ONLY)))
+        .thenReturn(true);
+    when(worldStateArchive.getMutable(any(Hash.class), any(Hash.class)))
+        .thenReturn(Optional.of(mock(MutableWorldState.class)));
+    when(blockProcessor.processBlock(eq(blockchain), any(MutableWorldState.class), eq(badBlock)))
+        .thenReturn(
+            new BlockProcessor.Result() {
+              @SuppressWarnings("unchecked")
+              @Override
+              public List<TransactionReceipt> getReceipts() {
+                return Collections.EMPTY_LIST;
+              }
+
+              @SuppressWarnings("unchecked")
+              @Override
+              public List<TransactionReceipt> getPrivateReceipts() {
+                return Collections.EMPTY_LIST;
+              }
+
+              @Override
+              public boolean isSuccessful() {
+                return false;
+              }
+
+              @Override
+              public Optional<Throwable> causedBy() {
+                return Optional.of(new StorageException("database bedlam"));
               }
             });
     when(blockBodyValidator.validateBody(

